universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java

```diff
@@ -0,0 +1,169 @@
+            + 1;
+    BlockDefinition blockDefinition =
+        BlockDefinition.builder()
+            .setId(blockId)
+            .setName(blockName)
+            .setDescription(blockDescription)
+            .build();
+    Program program =
+        programDefinition.toBuilder().addBlockDefinition(blockDefinition).build().toProgram();
+    return programRepository.updateProgramSync(program).getProgramDefinition();
+  }
+
+  @Override
+  @Transactional
+  public ProgramDefinition setBlockQuestions(
+      long programId, long blockDefinitionId, ImmutableList<QuestionDefinition> questionDefinitions)
+      throws ProgramNotFoundException, ProgramBlockNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setQuestionDefinitions(questionDefinitions)
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  @Override
+  @Transactional
+  public ProgramDefinition setBlockHidePredicate(
+      long programId, long blockDefinitionId, Predicate predicate)
+      throws ProgramNotFoundException, ProgramBlockNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setHidePredicate(Optional.of(predicate))
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  @Override
+  @Transactional
+  public ProgramDefinition setBlockOptionalPredicate(
+      long programId, long blockDefinitionId, Predicate predicate)
+      throws ProgramNotFoundException, ProgramBlockNotFoundException {
+    ProgramDefinition programDefinition =
+        getProgramDefinition(programId).orElseThrow(() -> new ProgramNotFoundException(programId));
+    int blockDefinitionIndex = getBlockDefinitionIndex(programDefinition, blockDefinitionId);
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setOptionalPredicate(Optional.of(predicate))
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
```

Comment:
nit: for the methods between L61 and here would you please add some blank lines in the method body to separate some of the statements? It'll make it a lot easier for me to read.

Reply:
done!